pgx/src/heap_tuple.rs

```diff
@@ -88,31 +90,68 @@ impl<'a> PgHeapTuple<'a, AllocatedByPostgres> {
         Self { tuple: PgBox::from_pg(heap_tuple), tupdesc }
     }
 
-    /// Creates a new [PgHeapTuple] from one of the two (`Current` or `New`) trigger tuples.  The returned
+    /// Creates a new [PgHeapTuple] identified by the `which_tuple` trigger tuple.  The returned
     /// [PgHeapTuple] will be considered by have been allocated by Postgres and is not mutable until
     /// [PgHeapTuple::into_owned] is called.
     ///
+    /// pgx also invents the concept of a "current" ([`TriggerTuple::Current`]) tuple, which is either
+    /// the new row being inserted or the row being updated (not the new version) or deleted.
+    ///
+    /// Asking for a [`TriggerTuple`] that isn't compatible with how the trigger was fired causes
+    /// this function to return `None`.  Specifically this means `None` is always returned for
+    /// statement-level triggers.
+    ///
     /// ## Safety
     ///
     /// This function is unsafe as we cannot guarantee that any pointers in the `trigger_data`
-    /// argument are valid.
+    /// argument are valid or that it's being used in the context of a firing trigger, which necessitates
+    /// Postgres internal state be correct for executing a trigger.
     pub unsafe fn from_trigger_data(
         trigger_data: &'a pg_sys::TriggerData,
         which_tuple: TriggerTuple,
     ) -> Option<PgHeapTuple<'a, AllocatedByPostgres>> {
-        let tupdesc =
-            PgTupleDesc::from_pg_unchecked(trigger_data.tg_relation.as_ref().unwrap().rd_att);
+        if trigger_fired_for_statement(trigger_data.tg_event) {
+            // there is no HeapTuple for a statement-level trigger as such triggers aren't run
+            // per-row
+            return None;
+        }
 
         let tuple = match which_tuple {
             TriggerTuple::Current => trigger_data.tg_trigtuple,
-            TriggerTuple::New => trigger_data.tg_newtuple,
+            TriggerTuple::New => {
+                if trigger_fired_by_insert(trigger_data.tg_event) {
+                    trigger_data.tg_trigtuple
+                } else if trigger_fired_by_update(trigger_data.tg_event) {
+                    trigger_data.tg_newtuple
+                } else {
+                    return None;
+                }
+            }
+            TriggerTuple::Old => {
+                if trigger_fired_by_update(trigger_data.tg_event)
+                    || trigger_fired_by_delete(trigger_data.tg_event)
+                {
+                    trigger_data.tg_trigtuple
+                } else {
+                    return None;
+                }
+            }
         };
 
-        if tuple.is_null() {
-            return None;
-        }
+        // at this point we should not be trying to return a NULL trigger tuple.
+        // We should have done that above as an early return
+        assert!(!tuple.is_null());
```

Comment:
What logic path do we have that could possibly have reached this point?

Reply:
This was just a guard against the `trigger_data.tg_trigtuple/tg_newtuple` fields being NULL when we should have the proper logic in place to ensure that won't be possible, based on how the trigger function is being called from Postgres.

I'll expand the comment.

Reply:
See if this is any better:  https://github.com/tcdi/pgx/commit/0f840d23cc9db2f1334711cf6e8a89f84b9e9df6